File: saucelabs-selenium-project/utils/config_reader.py
"""
config_reader.py - Reads configuration from config.ini and .env files.
"""
import configparser
import os
from dotenv import load_dotenv

# Load environment variables from .env
load_dotenv()


class ConfigReader:
    """Singleton configuration reader for the test framework."""

    _config = None

    @classmethod
    def _load_config(cls):
        if cls._config is None:
            cls._config = configparser.ConfigParser()
            config_path = os.path.join(os.path.dirname(os.path.dirname(__file__)), "config.ini")
            cls._config.read(config_path)
        return cls._config

    @classmethod
    def get(cls, section, key, fallback=None):
        """Get a config value by section and key."""
        config = cls._load_config()
        return config.get(section, key, fallback=fallback)

    @classmethod
    def get_base_url(cls):
        return cls.get("application", "base_url")

    @classmethod
    def get_browser(cls):
        return cls.get("browser", "name", fallback="chrome")

    @classmethod
    def is_headless(cls):
        return cls.get("browser", "headless", fallback="false").lower() == "true"

    @classmethod
    def get_implicit_wait(cls):
        return int(cls.get("timeouts", "implicit_wait", fallback="10"))

    @classmethod
    def get_explicit_wait(cls):
        return int(cls.get("timeouts", "explicit_wait", fallback="15"))

    @classmethod
    def get_page_load_timeout(cls):
        return int(cls.get("timeouts", "page_load_timeout", fallback="30"))

    @classmethod
    def get_valid_username(cls):
        return cls.get("test_data", "valid_username")

    @classmethod
    def get_valid_password(cls):
        return cls.get("test_data", "valid_password")

    @classmethod
    def get_locked_username(cls):
        return cls.get("test_data", "locked_username")

    @classmethod
    def get_problem_username(cls):
        return cls.get("test_data", "problem_username")

    @classmethod
    def get_execution_mode(cls):
        return os.getenv("EXECUTION_MODE", "local").lower()

    @classmethod
    def get_sauce_username(cls):
        return os.getenv("SAUCE_USERNAME", "")

    @classmethod
    def get_sauce_access_key(cls):
        return os.getenv("SAUCE_ACCESS_KEY", "")

    @classmethod
    def is_sauce_enabled(cls):
        return cls.get_execution_mode() == "sauce"

File: src/main/java/com/saucelabs/config/ConfigReader.java
package com.saucelabs.config;

import io.github.cdimascio.dotenv.Dotenv;

import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Properties;

/**
 * ConfigReader - Reads configuration from config.properties and .env files.
 * Singleton pattern ensures config is loaded only once.
 */
public class ConfigReader {

    private static Properties properties;
    private static Dotenv dotenv;

    static {
        loadConfig();
        loadDotenv();
    }

    private static void loadConfig() {
        properties = new Properties();
        try (InputStream input = ConfigReader.class.getClassLoader()
                .getResourceAsStream("config.properties")) {
            if (input != null) {
                properties.load(input);
            } else {
                throw new RuntimeException("config.properties not found in resources");
            }
        } catch (IOException e) {
            throw new RuntimeException("Failed to load config.properties: " + e.getMessage());
        }
    }

    private static void loadDotenv() {
        try {
            dotenv = Dotenv.configure().ignoreIfMissing().load();
        } catch (Exception e) {
            dotenv = null;
        }
    }

    private static String getEnv(String key) {
        String val = System.getProperty(key);
        if (val != null && !val.isEmpty()) return val;
        if (dotenv != null) {
            val = dotenv.get(key);
            if (val != null && !val.isEmpty()) return val;
        }
        return System.getenv(key);
    }

    public static String get(String key) {
        return properties.getProperty(key);
    }

    public static String get(String key, String fallback) {
        return properties.getProperty(key, fallback);
    }

    // ---- Application ----

    public static String getBaseUrl() {
        return get("application.base_url");
    }

    // ---- Browser ----

    public static String getBrowser() {
        return get("browser.name", "chrome");
    }

    public static boolean isHeadless() {
        return Boolean.parseBoolean(get("browser.headless", "false"));
    }

    // ---- Timeouts ----

    public static int getImplicitWait() {
        return Integer.parseInt(get("timeouts.implicit_wait", "10"));
    }

    public static int getExplicitWait() {
        return Integer.parseInt(get("timeouts.explicit_wait", "15"));
    }

    public static int getPageLoadTimeout() {
        return Integer.parseInt(get("timeouts.page_load_timeout", "30"));
    }

    // ---- Test Data ----

    public static String getValidUsername() {
        return get("test_data.valid_username");
    }

    public static String getValidPassword() {
        return get("test_data.valid_password");
    }

    public static String getLockedUsername() {
        return get("test_data.locked_username");
    }

    public static String getProblemUsername() {
        return get("test_data.problem_username");
    }

    // ---- Execution Mode ----

    public static String getExecutionMode() {
        String env = getEnv("EXECUTION_MODE");
        return (env != null && !env.isEmpty()) ? env.toLowerCase() : "local";
    }

    public static boolean isSauceEnabled() {
        return "sauce".equals(getExecutionMode());
    }

    // ---- Sauce Labs ----

    public static String getSauceUsername() {
        String val = getEnv("SAUCE_USERNAME");
        return (val != null) ? val : "";
    }

    public static String getSauceAccessKey() {
        String val = getEnv("SAUCE_ACCESS_KEY");
        return (val != null) ? val : "";
    }

    public static String getSauceBrowserName() {
        return get("saucelabs.browser_name", "chrome");
    }

    public static String getSauceBrowserVersion() {
        return get("saucelabs.browser_version", "latest");
    }

    public static String getSaucePlatformName() {
        return get("saucelabs.platform_name", "Windows 11");
    }

    public static String getSauceBuildName() {
        return get("saucelabs.build_name", "Local Build");
    }

    public static String getSauceScreenResolution() {
        return get("saucelabs.screen_resolution", "1920x1080");
    }

    public static String getSauceUrl() {
        return get("saucelabs.url",
                "https://ondemand.us-west-1.saucelabs.com:443/wd/hub");
    }
}
